### PR TITLE
[AMDGPU] modify named barrier builtins and intrinsics

### PR DESCRIPTION
**Pull request by cmc-rep**
Use a local pointer type to represent the named barrier in builtin and intrinsic. This makes the definitions more user friendly 
bacause they do not need to worry about the hardware ID assignment. Also this approach is more like the other popular GPU programming language.  Named barriers should be represented as global variables of addrspace(3) in LLVM-IR. Compiler assigns the special LDS offsets for those variables during AMDGPULowerModuleLDS pass. Those addresses are converted to hw barrier ID during instruction selection. The rest of the instruction-selection changes are primarily due to the intrinsic-definition changes.

### CODE REVIEW

File: llvm/lib/Target/AMDGPU/AMDGPULowerModuleLDSPass.cpp [hunk 1]
```diff
@@ -920,6 +920,124 @@ class AMDGPULowerModuleLDS {
     return KernelToCreatedDynamicLDS;
   }
 
+  static GlobalVariable *uniquifyGVPerKernel(Module &M, GlobalVariable *GV,
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/AMDGPULowerModuleLDSPass.cpp:923]: This looks like an unrelated, separate change

**Review by arsenm — COMMENTED** [review]
Modify how? There seem to be too many things going on here. Description should say how

File: llvm/lib/Target/AMDGPU/AMDGPULowerModuleLDSPass.cpp (hunk 1 above)
Comment (cmc-rep) [reply llvm/lib/Target/AMDGPU/AMDGPULowerModuleLDSPass.cpp:923]: I will add you to the internal review link. So you can see the review history.

Comment (arsenm) [reply llvm/lib/Target/AMDGPU/AMDGPULowerModuleLDSPass.cpp:923]: This still should be a separate PR

Comment (cmc-rep) [reply llvm/lib/Target/AMDGPU/AMDGPULowerModuleLDSPass.cpp:923]: This is where the local pointer address is assigned for a named barrier before going into AMDGPUbackend

Comment (cmc-rep) [reply llvm/lib/Target/AMDGPU/AMDGPULowerModuleLDSPass.cpp:923]: This goes together with the change in AMDGPUMachineFunction.cpp

File: llvm/lib/Target/AMDGPU/AMDGPUMachineFunction.cpp [hunk 2]
```diff
@@ -102,6 +103,12 @@ unsigned AMDGPUMachineFunction::allocateLDSGlobal(const DataLayout &DL,
 
   unsigned Offset;
   if (GV.getAddressSpace() == AMDGPUAS::LOCAL_ADDRESS) {
+    if (TargetExtType *TTy = AMDGPU::isNamedBarrier(GV)) {
+      auto Offset = getLDSAbsoluteAddress(GV);
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/AMDGPUMachineFunction.cpp:107]: no auto

File: llvm/lib/Target/AMDGPU/AMDGPUMachineFunction.cpp [hunk 3]
```diff
@@ -102,6 +103,12 @@ unsigned AMDGPUMachineFunction::allocateLDSGlobal(const DataLayout &DL,
...
+      assert(Offset && "named barrier should have an absolute address");
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/AMDGPUMachineFunction.cpp:108]: This doesn't seem like something appropriate to assert on. Is there a verifier check for this?

File: llvm/lib/Target/AMDGPU/AMDGPUMemoryUtils.cpp [hunk 4]
```diff
@@ -32,6 +32,28 @@ Align getAlign(const DataLayout &DL, const GlobalVariable *GV) {
...
+  // - Allocas.
+  Type *Ty = GV.getValueType();
+  while (true) {
+    if (auto *TTy = dyn_cast<TargetExtType>(Ty))
+      return TTy->getName() == "amdgcn.named.barrier" ? TTy : nullptr;
+    if (auto *STy = dyn_cast<StructType>(Ty)) {
+      if (STy->getNumElements() == 0)
+        return nullptr;
+      Ty = STy->getElementType(0);
+      continue;
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/AMDGPUMemoryUtils.cpp:51]: Don't understand why this needs to look through nested structs or uses an infinite loop

File: llvm/lib/Target/AMDGPU/AMDGPULowerModuleLDSPass.cpp (hunk 1 above)
Comment (arsenm) [reply llvm/lib/Target/AMDGPU/AMDGPULowerModuleLDSPass.cpp:923]: The backend and frontend parts should still be split. Description also does not say how or what the modification is

**Comment by cmc-rep** [comment]
Because  I am changing the existing builtins and intrinsics, it is very difficult to split the patch into multiple parts, and still make it sure the modified piece connects with the not-modified pieces.  I am really not sure how to do that

**Comment by cmc-rep** [comment]
Also, I added the piece in AMDGPULowerModuleLDS during internal review because other reviewers want to see the functional completeness.

File: llvm/lib/Target/AMDGPU/AMDGPUMachineFunction.cpp (hunk 3 above)
Comment (cmc-rep) [reply llvm/lib/Target/AMDGPU/AMDGPUMachineFunction.cpp]: Changed it.

**Review by rampitec — APPROVED** [review]
It's the same code already reviewed downstream. LGTM.

**Comment by arsenm** [comment]
> It's the same code already reviewed downstream. LGTM.

The commit message is still not helpful

**Comment by JonChesterfield** [comment]
Uh, why? In what sense are named barriers LDS globals?

**Comment by Pierre-vh** [comment]
> Uh, why? In what sense are named barriers LDS globals?

I also disagree with this direction and I personally would like to see this reverted, or at least fixed in a new patch, before anything else lands on top.

First, this patch just does way too much in one go. The new barrier ID system + the other intrinsic changes are separate patches, IMO.

These are not LDS global, and the fact that we "theoretically could lower them as LDS operation" is really not a valid argument.
We could also lower them as global operations, scratch operations, or any other system we want really. 
I also reject the idea of lowering a `s.barrier` (**scalar** barrier) as a set of LDS operations with different properties. That intrinsic has the naming convention of intrinsics that directly map to one hardware instruction.

Lastly this has no documentation whatsoever to explain how to iterate on the design in the future.

I agree it's a good idea to have another system than raw IDs for the named barriers. Adding fail-safe lowerings is also a good idea, but it has to be done differently. If we want to have a generic barrier that can have "universal but slow" lowerings like this, we should have a separate, generic barrier intrinsic (e.g. a `amdgpu.execution.barrier` intrinsic). If this is desired I can look into it because I've been working on a barrier execution model for the past few months.

I think a better way to approach this is to follow a pattern similar to `amdgpu.kernel.runtime.handle` (cc @arsenm )

- Just use the generic AS.
- Make the global `externally_initialized`
- Maybe make it a constant as well. The ID is like a pointer to the barrier state, and the ID never changes. Only the state does.
- Leverage the `section` attribute, e.g. put it in `.amdgpu.workgroup_barrier`
  - This describes where it lives on the HW, we can document it as the "An opaque, implementation-defined barrier state stored at workgroup scope in an unspecified memory region".
- Keep using the target type with something like `amdgpu.barrier_id`

Then we have a separate, minimal pass that lowers these and eliminates the globals before ISel.
Finally, add documentation on top to clearly define the intended semantics so anyone that wants to iterate on this system later can do it safely.

I can help with this as I've been working on barriers for a few months now. Feel free to reach out.

File: llvm/lib/Target/AMDGPU/SIISelLowering.cpp [hunk 5]
```diff
@@ -9386,27 +9386,33 @@ SDValue SITargetLowering::LowerINTRINSIC_W_CHAIN(SDValue Op,
...
       Opc = AMDGPU::S_GET_BARRIER_STATE_M0;
-      SDValue M0Val = copyToM0(DAG, Chain, DL, Op.getOperand(2));
-      Ops.push_back(M0Val.getValue(0));
+      if (IntrID == Intrinsic::amdgcn_s_get_named_barrier_state) {
+        SDValue M0Val;
+        M0Val = DAG.getNode(ISD::SRL, DL, MVT::i32, Op->getOperand(2),
+                            DAG.getShiftAmountConstant(4, MVT::i32, DL));
+        M0Val = SDValue(
+            DAG.getMachineNode(AMDGPU::S_AND_B32, DL, MVT::i32, M0Val,
+                               DAG.getTargetConstant(0x3F, DL, MVT::i32)),
```
Comment (easyonaadit) [inline llvm/lib/Target/AMDGPU/SIISelLowering.cpp:9411]: Hi, Are you sure that `0x3F` is the right constant?
Ad defined in `SIDefines.h`:
```
  NAMED_BARRIER_FIRST = 1,
  NAMED_BARRIER_LAST = 16,
```
That corresponds to 5 bits, so shouldn't it be `0x1F`?

File: llvm/lib/Target/AMDGPU/AMDGPUInstructionSelector.cpp [hunk 6]
```diff
@@ -5457,110 +5448,168 @@ bool AMDGPUInstructionSelector::selectSBarrierSignalIsfirst(
...
+  Register TmpReg0 = MRI->createVirtualRegister(&AMDGPU::SReg_32RegClass);
+  BuildMI(*MBB, &I, DL, TII.get(AMDGPU::S_LSHR_B32), TmpReg0)
+      .add(BarOp)
+      .addImm(4u)
+      .setOperandDead(3); // Dead scc
+
+  Register TmpReg1 = MRI->createVirtualRegister(&AMDGPU::SReg_32RegClass);
+  BuildMI(*MBB, &I, DL, TII.get(AMDGPU::S_AND_B32), TmpReg1)
+      .addReg(TmpReg0)
+      .addImm(0x3F)
```
Comment (easyonaadit) [inline llvm/lib/Target/AMDGPU/AMDGPUInstructionSelector.cpp:5525]: Ditto.
